File: 002 addTwoNumbers/002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode],carry=0) -> Optional[ListNode]:
        # 递归法
        if l1 is None and l2 is None:
            return ListNode(carry)if carry else None
        if l1 is None:
            l1,l2=l2,l1
        carry+= l1.val+(l2.val if l2 else 0)
        l1.val=carry%10
        l1.next=self.addTwoNumbers(l1.next,l2.next if l2 else None,carry//10)
        return l1

File: 002 addTwoNumbers/002.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    struct ListNode *head=NULL,*tail=NULL;
    int carry=0;
    while(l1 || l2){
        int n1=l1?l1->val:0;
        int n2=l2?l2->val:0;
        int sum=n1+n2+carry;
        if(!head){
            head=(struct ListNode*)malloc(sizeof(struct ListNode));
            tail=head;
            tail->val=sum%10;
            tail->next=NULL;
        }else{
            tail->next=(struct ListNode*)malloc(sizeof(struct ListNode));
            tail=tail->next;
            tail->val=sum%10;
            tail->next=NULL;
        }
        carry=sum/10;
        if(l1){
            l1=l1->next;
        }
        if(l2){
            l2=l2->next;
        }
    }
    if(carry>0){
        tail->next=(struct ListNode*)malloc(sizeof(struct ListNode));
        tail=tail->next;
        tail->val=carry;
        tail->next=NULL;
    }
    return head;
}
